optuna/_gp/optim_mixed.py

```diff
@@ -0,0 +1,320 @@
+        negative_fval_cache[i] = negval
+        return negval
+
+    def interpolated_negative_acqf(x: float) -> float:
+        if x < grids[0] or x > grids[-1]:
+            return np.inf
+        right = int(np.clip(np.searchsorted(grids, x), 1, len(grids) - 1))
+        left = right - 1
+        neg_acqf_left, neg_acqf_right = negative_acqf_with_cache(left), negative_acqf_with_cache(
+            right
+        )
+        w_left = (grids[right] - x) / (grids[right] - grids[left])
+        w_right = 1.0 - w_left
+        return w_left * neg_acqf_left + w_right * neg_acqf_right
+
+    EPS = 1e-12
+    res = so.minimize_scalar(
+        interpolated_negative_acqf,
+        # The values of this bracket are (inf, -fval, inf).
+        # This trivially satisfies the bracket condition if fval is finite.
+        bracket=(grids[0] - EPS, grids[current_choice_i], grids[-1] + EPS),
+        method="brent",
+        tol=xtol,
+    )
+    opt_idx = find_nearest_index(res.x)
+    fval_opt = -negative_acqf_with_cache(opt_idx)
+
+    # We check both conditions because of numerical errors.
+    if opt_idx != current_choice_i and fval_opt > initial_fval:
+        normalized_params[param_idx] = grids[opt_idx]
+        return (normalized_params, fval_opt, True)
+    else:
+        return (initial_params, initial_fval, False)
+
+
+def _local_search_discrete(
+    acqf_params: AcquisitionFunctionParams,
+    initial_params: np.ndarray,
+    initial_fval: float,
+    param_idx: int,
+    choices: np.ndarray,
+    xtol: float,
+) -> tuple[np.ndarray, float, bool]:
+
+    # If the number of possible parameter values is small, we just perform an exhaustive search.
+    # This is faster and better than the line search.
+    MAX_INT_EXHAUSTIVE_SEARCH_PARAMS = 16
+
+    scale_type = acqf_params.search_space.scale_types[param_idx]
+    if scale_type == ScaleType.CATEGORICAL or len(choices) <= MAX_INT_EXHAUSTIVE_SEARCH_PARAMS:
+        return _exhaustive_search(acqf_params, initial_params, initial_fval, param_idx, choices)
+    else:
+        return _discrete_line_search(
+            acqf_params, initial_params, initial_fval, param_idx, choices, xtol
+        )
+
+
+def local_search_mixed(
+    acqf_params: AcquisitionFunctionParams,
+    initial_normalized_params: np.ndarray,
+    *,
+    tol: float = 1e-4,
+    max_iter: int = 100,
+) -> tuple[np.ndarray, float]:
+    scale_types = acqf_params.search_space.scale_types
+    bounds = acqf_params.search_space.bounds
+    steps = acqf_params.search_space.steps
+
+    continuous_indices = np.where(steps == 0.0)[0]
+
+    inverse_squared_lengthscales = (
+        acqf_params.kernel_params.inverse_squared_lengthscales.detach().numpy()
+    )
+    # This is a technique for speeding up optimization.
+    # We use an isotropic kernel, so scaling the gradient will make
+    # the hessian better-conditioned.
+    lengthscale = 1 / np.sqrt(inverse_squared_lengthscales[continuous_indices])
```

Comment:
`inverse_squared_lengthscales` is plural, so I also changed it to the plural form. 
```suggestion
    lengthscales = 1 / np.sqrt(inverse_squared_lengthscales[continuous_indices])
```